File: Problems/0373-find-k-pairs-with-smallest-sums.py
import heapq

# LeetCode solution starts from here
class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        if not nums1 or not nums2 or k == 0:
            return []

        heap = []

        for i in range(min(len(nums1), k)):
            heapq.heappush(heap, (nums1[i] + nums2[0], i, 0))

        ans = []

        while heap and k:
            _, i, j = heapq.heappop(heap)
            ans.append([nums1[i], nums2[j]])
            k -= 1

            if j + 1 < len(nums2):
                heapq.heappush(heap, (nums1[i] + nums2[j + 1], i, j + 1))

        return ans

File: Problems/0373-find-k-pairs-with-smallest-sums.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here
class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        if (nums1.empty() || nums2.empty() || k == 0)
            return {};

        using T = tuple<int, int, int>;
        priority_queue<T, vector<T>, greater<T>> heap;

        for (int i = 0; i < min((int)nums1.size(), k); i++)
            heap.emplace(nums1[i] + nums2[0], i, 0);

        vector<vector<int>> ans;

        while (!heap.empty() && k--) {
            auto [sum, i, j] = heap.top();
            heap.pop();

            ans.push_back({nums1[i], nums2[j]});

            if (j + 1 < nums2.size())
                heap.emplace(nums1[i] + nums2[j + 1], i, j + 1);
        }

        return ans;
    }
};